File: algorithms/leetcode/005-longest-palindromic-substring/longest-palindromic-substring.js
const longestPalindrome = s => {
    let maxLen = 1;
    let maxStr = s[0];

    const findPalindrome = (begin, end, curLen) => {
        while(begin > -1 && end < s.length && s[begin] === s[end]) {
            curLen += 2;

            if (curLen > maxLen) {
                maxLen = curLen;
                maxStr = s.substring(begin, end + 1);
            }

            begin--;
            end++;
        }
    };

    for (let i = 0; i < s.length; i++) {
        findPalindrome(i - 1, i + 1, 1);
        findPalindrome(i, i + 1, 0);
    }

    return maxStr;
};

File: algorithms/leetcode/005-longest-palindromic-substring/LongestPalindromicSubstring.java
class Solution {
    int maxLength = 0;
    String maxString = "";
    
    private void scan(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            int length = right - left + 1;
            if (length > maxLength) {
                maxLength = length;
                maxString = s.substring(left, right + 1);
            }
            left--;
            right++;
        }
    }
    
    public String longestPalindrome(String s) {
        if (s.length() > 0) {
            maxString = String.valueOf(s.charAt(0));
            
            for (int i = 0; i < s.length(); i++) {
                scan(s, i - 1, i + 1);
                scan(s, i, i + 1);
            }
        }
        
        return maxString;
    }
}
